src/ssl/gadgets.cc

```diff
@@ -479,16 +479,28 @@ static bool buildCertificate(Security::CertPointer & cert, Ssl::CertificatePrope
         (void)replaceCommonName(cert, properties.commonName);
     }
 
-    // We should get caCert notBefore and notAfter fields and do not allow
-    // notBefore/notAfter values from certToMimic before/after notBefore/notAfter
-    // fields from caCert.
-    // Currently there is not any way in openssl tollkit to compare two ASN1_TIME
-    // objects.
+    ASN1_TIME *caNotBefore = nullptr;
+    ASN1_TIME *caNotAfter = nullptr;
+
+    // Get times from cacert
+    if (const auto caCert = properties.signWithX509.get()) {
+        caNotBefore = X509_getm_notBefore(caCert);
+        caNotAfter = X509_getm_notAfter(caCert);
+    }
+
     ASN1_TIME *aTime = NULL;
-    if (!properties.setValidBefore && properties.mimicCert.get())
+    if (!properties.setValidBefore && properties.mimicCert.get()) {
         aTime = X509_getm_notBefore(properties.mimicCert.get());
-    if (!aTime && properties.signWithX509.get())
-        aTime = X509_getm_notBefore(properties.signWithX509.get());
+
+#if HAVE_LIBCRYPTO_ASN1_TIME_COMPARE
+        // If time is before cacert, fix it
+        if (caNotBefore && ASN1_TIME_compare(aTime, caNotBefore) == -1)
```

Comment:
AFAICT by looking at OpenSSL sources, ASN1_TIME_compare() will do strange, unwanted things when aTime is nil. The old code implied that aTime can be nil here. I suggest adding a simple static ASN1_TIME_compare() wrapper that will return -2 when at least one of its ASN1_TIME pointer arguments is nil and use that wrapper here:

```suggestion
        if (xASN1_TIME_compare(aTime, caNotBefore) == -1)
```

As an added bonus, you will move `HAVE_LIBCRYPTO_ASN1_TIME_COMPARE` inside that wrapper, reducing code duplication:

```
/// ASN1_TIME_compare() wrapper that
/// \returns -2 when one of the arguments is nil or ASN1_TIME_compare() is missing
int
ASN1_TIME_compare(ASN1_TIME *a, ASN1_TIME *b)
{
    if (!a || !b)
        return -2;
#if HAVE_LIBCRYPTO_ASN1_TIME_COMPARE
    return ASN1_TIME_compare(a, b); // also returns -2 on errors
#else  
    return -2; // OpenSSL v1.0 lacks ASN1_TIME_compare()
#endif
}
```

The `if` statement in the wrapper is left outside HAVE_LIBCRYPTO_ASN1_TIME_COMPARE check on purpose -- to avoid unused function parameter warnings from the compiler.

The wrapper can be added right above the buildCertificate() function you are editing. As a static helper function, it does not need to go into any header file.

Reply:
Integrated! Thanks :-)